boa/src/syntax/parser/function/mod.rs

```diff
@@ -188,10 +201,64 @@ where
         let _timer = BoaProfiler::global().start_event("BindingRestElement", "Parsing");
         cursor.expect(Punctuator::Spread, "rest parameter")?;
 
-        let param = BindingIdentifier::new(self.allow_yield, self.allow_await).parse(cursor)?;
-        // TODO: BindingPattern
+        if let Some(t) = cursor.peek(0)? {
+            let declaration = match *t.kind() {
+                TokenKind::Punctuator(Punctuator::OpenBlock) => {
+                    let param = ObjectBindingPattern::new(true, self.allow_yield, self.allow_await)
+                        .parse(cursor)?;
+
+                    let init = if let Some(t) = cursor.peek(0)? {
+                        // Check that this is an initilizer before attempting parse.
+                        if *t.kind() == TokenKind::Punctuator(Punctuator::Assign) {
+                            Some(
+                                Initializer::new(true, self.allow_yield, self.allow_await)
+                                    .parse(cursor)?,
+                            )
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    };
```

Comment:
This should be simpler with a functional approach:

```suggestion
                    let init = cursor.peek(0)?.filter(|t| {
                        // Check that this is an initilizer before attempting parse.
                        *t.kind() == TokenKind::Punctuator(Punctuator::Assign)
                    }).and_then(|t|
                        Some(Initializer::new(true, self.allow_yield, self.allow_await).parse(cursor))
                    ).transpose()?;
```

Applies to all patterns similar to this.

Reply:
@jedel1043 
above changes gives errors because borrow occurs at `cursor.peek()` and `*cursor` is needed in the same closure,
should I implement below code using `.cloned` or is there an efficient way to do it?
```
let init = cursor.peek(0)?.cloned().filter(|t| {
                        // Check that this is an initializer before attempting parse.
                        *t.kind() == TokenKind::Punctuator(Punctuator::Assign)
                    }).map(|_|
                        Initializer::new(true, self.allow_yield, self.allow_await).parse(cursor)
                    ).transpose()?;
```

Reply:
```Rust
let init = cursor.peek(0)?;
let init = init.filter(..)
...
```

Reply:
same problem still exists

Reply:
Then the problem must be `map`. Hmm... does this work?

```Rust
let init = cursor.peek(0)?.filter(|t| { // Check that this is an initilizer before attempting parse. *t.kind() == TokenKind::Punctuator(Punctuator::Assign) }).map(|t| {
drop(t);
Initializer::new(true, self.allow_yield, self.allow_await).parse(cursor)
}).transpose()?;
```

If it doesn't, just clone. I prefer focusing on the readability of the code, we can optimize later.

Reply:
filter expects a bool
```
expected type `bool`
               found enum `Result<Node, parser::error::ParseError>`
```

Reply:
I reedited the code, had a missing \n

Reply:
the original error again